packages/data/src/components/use-select/index.js

```diff
@@ -75,42 +57,50 @@ const renderQueue = createQueue();
  * any price in the state for that currency is retrieved. If the currency prop
  * doesn't change and other props are passed in that do change, the price will
  * not change because the dependency is just the currency.
- *
  * @return {Function}  A custom react hook.
  */
 export default function useSelect( _mapSelect, deps ) {
 	const mapSelect = useCallback( _mapSelect, deps );
+	const previousMapSelect = usePrevious( mapSelect );
+	const result = useRef();
 	const registry = useRegistry();
 	const isAsync = useAsyncMode();
-	// React can sometimes clear the `useMemo` cache.
-	// We use the cache-stable `useMemoOne` to avoid
-	// losing queues.
-	const queueContext = useMemoOne( () => ( { queue: true } ), [ registry ] );
-	const [ , forceRender ] = useReducer( ( s ) => s + 1, 0 );
-
-	const latestMapSelect = useRef();
-	const latestIsAsync = useRef( isAsync );
-	const latestMapOutput = useRef();
-	const latestMapOutputError = useRef();
-	const isMountedAndNotUnsubscribing = useRef();
-
-	let mapOutput;
+	const [ , dispatch ] = useState( {} );
+	const rerender = () => dispatch( {} );
+	const isMountedAndNotUnsubscribing = useRef( true );
```

Comment:
There is a package called [use-subcription](https://www.npmjs.com/package/use-subscription), directly from the React team, that provides a `useSubscription` hook for subscribing to things from React components in async mode.

It solves all the gotchas and race conditions that need to be taken care of. Using it in `useSelect` (both in this new and the legacy version) would simplify the hook tremendously, outsourcing most of the complexity. All that remains is calling `mapSelect` and memoizing it properly, and handling the "async mode" where the state updates are fired through the priority queue rather than done directly.

Reply:
Oh good reminder, I actually did try that package once of useSelect but at that time, it was causing a big loss in performance, maybe with the atoms implementation, it will be better.